main/src/library/Fixpoint3/Interpreter.flix

```diff
@@ -0,0 +1,699 @@
+        evalOp(rc, ctxt, parLevel, body)
+
+    ///
+    /// Handles executing `body` for a `Search(Latticenal)` or `Query(Latticenal)`
+    /// RelOp with current bound tuple `t` and lattice value `l`.
+    ///
+    /// Should only be used when new threads are being spawned.
+    ///
+    /// `body` is the body of the `Search` or `Query`. `writes` is a vector of writes to `SearchTuple`'s
+    /// to be performed. See `handleWriteTuples`.
+    ///
+    def parLatticeHandlingBodyFunc(
+        rc: Region[r],
+        tuplePos: Int32,
+        meetWith: Int32,
+        glb: Boxed -> Boxed -> Boxed,
+        leq: Boxed -> Boxed -> Bool,
+        bot: Boxed,
+        writes: Vector[WriteTuple],
+        body: RelOp,
+        ctxt: Context[r],
+        parLevel: AllowedBranches,
+        t: Tuple,
+        l: Boxed
+    ): Unit \ r =
+        latticeHandlingBodyFunc(rc, tuplePos, meetWith, glb, leq, bot, writes, body, makeCtxt(rc, ctxt), parLevel, t, l)
+
+    ///
+    /// Handles executing `body` for a `Search(Latticenal)` or `Query(Latticenal)`
+    /// RelOp with current bound tuple `t` and lattice value `l`.
+    ///
+    /// Assumes exclusive access to `ctxt`.
+    ///
+    /// `body` is the body of the `Search` or `Query`. `writes` is a vector of writes to `SearchTuple`'s
+    /// to be performed. See `handleWriteTuples`.
+    ///
+    def latticeHandlingBodyFunc(
+        rc: Region[r],
+        tuplePos: Int32,
+        meetWith: Int32,
+        glb: Boxed -> Boxed -> Boxed,
+        leq: Boxed -> Boxed -> Bool,
+        bot: Boxed,
+        writes: Vector[WriteTuple],
+        body: RelOp,
+        ctxt: Context[r],
+        parLevel: AllowedBranches,
+        t: Tuple,
+        l: Boxed
+    ): Unit \ r =
+        let (_, env, _) = ctxt;
+        let (minEnv, maxEnv, tupleEnv, latEnv) = env;
+        handleWriteTuples(t, writes, minEnv, maxEnv, tuplePos, tupleEnv);
+        // Recurse only if the meet is not bottom
+        if (meetWith == tuplePos) {
+            Array.put(l, tuplePos, latEnv);
+            evalOp(rc, ctxt, parLevel, body)
+        } else {
+            let meet = l `glb` Array.get(meetWith, latEnv);
+            if (meet `leq` bot) ()
+            else {
+                Array.put(meet, tuplePos, latEnv);
+                evalOp(rc, ctxt, parLevel, body)
+            }
+        }
+
+    ///
+    /// Evaluates `terms`, where the last value of `terms` will 
+    ///
+    def evalTermsWithLat(ctxt: Context[r], terms: Vector[RamTerm]): Tuple \ r =
+        let (_, env, boxing) = ctxt;
+        let evalT = evalTerm(env, boxing);
+        let len = Vector.length(terms);
+        Vector.range(0, len - 1) |> 
+        Vector.map(i -> evalT(Vector.get(i, terms)))
+
+    ///
+    /// Evaluate `exps` under context `ctxt` and returns true iff all `BoolExp` in `exps`
+    /// are true.
+    ///
+    def evalBoolExps(ctxt: Context[r], exps: Vector[BoolExp]): Bool \ r =
```

Comment:
Plural functions are unusual. I would have what you have, but both at the top-level:

evalBoolExps
evalBoolExp

then evalBoolExps just does the Vector.forAll-- clean and simple